pkg/dsn_util/postgresql.go

```diff
@@ -0,0 +1,452 @@
+	WithDSN(dsn)(c)
+	return
+}
+
+func (x *Config) ToDSN(isPasswordMosaic ...bool) string {
+	buff := strings.Builder{}
+	buff.WriteString("postgres://")
+	if x.User != "" || x.Password != "" {
+		buff.WriteString(x.User)
+		buff.WriteString(":")
+
+		if len(isPasswordMosaic) != 0 && isPasswordMosaic[0] {
+			buff.WriteString("*******")
+		} else {
+			buff.WriteString(x.Password)
+		}
+
+		buff.WriteString("@")
+	}
+	buff.WriteString(x.Addr)
+	if x.Database != "" {
+		buff.WriteString("/")
+		buff.WriteString(x.Database)
+	}
+	return buff.String()
+}
+
+type Option func(cfg *Config)
+
+// Deprecated. Use Option instead.
+type DriverOption = Option
+
+func WithNetwork(network string) Option {
+	if network == "" {
+		panic("network is empty")
+	}
+	return func(cfg *Config) {
+		cfg.Network = network
+	}
+}
+
+func WithAddr(addr string) Option {
+	if addr == "" {
+		panic("addr is empty")
+	}
+	return func(cfg *Config) {
+		cfg.Addr = addr
+	}
+}
+
+func WithTLSConfig(tlsConfig *tls.Config) Option {
+	return func(cfg *Config) {
+		cfg.TLSConfig = tlsConfig
+	}
+}
+
+func WithInsecure(on bool) Option {
+	return func(cfg *Config) {
+		if on {
+			cfg.TLSConfig = nil
+		} else {
+			cfg.TLSConfig = &tls.Config{InsecureSkipVerify: true}
+		}
+	}
+}
+
+func WithUser(user string) Option {
+	if user == "" {
+		panic("user is empty")
+	}
+	return func(cfg *Config) {
+		cfg.User = user
+	}
+}
+
+func WithPassword(password string) Option {
+	return func(cfg *Config) {
+		cfg.Password = password
+	}
+}
+
+func WithDatabase(database string) Option {
+	if database == "" {
+		panic("database is empty")
+	}
+	return func(cfg *Config) {
+		cfg.Database = database
+	}
+}
+
+func WithApplicationName(appName string) Option {
+	return func(cfg *Config) {
+		cfg.AppName = appName
+	}
+}
+
+func WithConnParams(params map[string]interface{}) Option {
+	return func(cfg *Config) {
+		cfg.ConnParams = params
+	}
+}
+
+func WithTimeout(timeout time.Duration) Option {
+	return func(cfg *Config) {
+		cfg.DialTimeout = timeout
+		cfg.ReadTimeout = timeout
+		cfg.WriteTimeout = timeout
+	}
+}
+
+func WithDialTimeout(dialTimeout time.Duration) Option {
+	return func(cfg *Config) {
+		cfg.DialTimeout = dialTimeout
+	}
+}
+
+func WithReadTimeout(readTimeout time.Duration) Option {
+	return func(cfg *Config) {
+		cfg.ReadTimeout = readTimeout
+	}
+}
+
+func WithWriteTimeout(writeTimeout time.Duration) Option {
+	return func(cfg *Config) {
+		cfg.WriteTimeout = writeTimeout
+	}
+}
+
+//// WithResetSessionFunc configures a function that is called prior to executing
+//// a query on a connection that has been used before.
+//// If the func returns driver.ErrBadConn, the connection is discarded.
+//func WithResetSessionFunc(fn func(context.Context, *Conn) error) Option {
+//	return func(cfg *Config) {
+//		cfg.ResetSessionFunc = fn
+//	}
+//}
+
+func WithDSN(dsn string) Option {
+	return func(cfg *Config) {
+		opts, err := parseDSN(dsn)
+		if err != nil {
+			panic(err)
```

Comment:
Oh, that's too bad.

Reply:
It's okay. There's a recovery at the entrance

Reply:
LGTM